core/primitives/src/child_trie.rs

```diff
@@ -0,0 +1,367 @@
+	res.extend_from_slice(prefix);
+	res.extend_from_slice(key.as_ref());
+	res
+}
+
+/// Parent trie origin. This type contains all information
+/// needed to access a parent trie.
+/// Currently only a single depth is supported for child trie,
+/// so it only contains the top trie key to the child trie.
+/// Internally this contains a full path key (with
+/// `well_known_keys::CHILD_STORAGE_KEY_PREFIX`).
+pub type ParentTrie = Vec<u8>;
+
+/// `ChildTrieReadRef` contains a reference to information
+/// needed to access a child trie content.
+/// Generally this should not be build directly but accessed
+/// through a `node_ref` function call.
+/// The struct can be build directly with invalid data, so
+/// its usage is limited to read only querying.
+#[derive(Clone)]
+pub enum ChildTrieReadRef<'a> {
+	/// Subtrie path for new trie, see [`ParentTrie`] for details.
+	New(&'a KeySpace),
+	/// Subtrie root hash and keyspace for existing child trie.
+	Existing(&'a [u8], &'a KeySpace),
+}
+
+impl<'a> ChildTrieReadRef<'a> {
+	/// Keyspace accessor for the enum.
+	pub fn keyspace(&self) -> &KeySpace {
+		match self {
+			ChildTrieReadRef::New(k) => k,
+			ChildTrieReadRef::Existing(_, k) => k,
+		}
+	}
+}
+/// Current codec version of a child trie definition.
+const LAST_SUBTRIE_CODEC_VERSION: u16 = 1u16;
+
+/// `ChildTrieNode` encoder internal implementation,
+/// it shall not be public.
+#[derive(Encode, Clone)]
+struct ChildTrieReadEncode<'a> {
+	/// Current codec version
+	#[codec(compact)]
+	version: u16,
+	/// Child trie unique keyspace
+	keyspace: &'a KeySpace,
+	/// Child trie root hash
+	root: &'a [u8],
+}
+
+#[derive(Decode)]
+struct ChildTrieReadDecode {
+	#[codec(compact)]
+	version: u16,
+	keyspace: KeySpace,
+	root: Vec<u8>,
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug, Hash, PartialOrd, Ord))]
+/// This struct contains information needed to access a child trie.
+/// When doing a remote query.
+pub struct ChildTrieRead {
+	/// Child trie parent key.
+	pub parent: ParentTrie,
+	/// Child trie root hash
+	pub root: Vec<u8>,
+}
+
+impl ChildTrieRead {
+	/// Use `ChildTrieRead` as a `ChildTrieReadRef`,
+	/// forcing root field to an existing value.
+	pub fn node_ref<'a>(&'a self, keyspace: &'a KeySpace) -> ChildTrieReadRef<'a> {
+		debug_assert!(self.root.len() > 0);
+		ChildTrieReadRef::Existing(&self.root[..], keyspace)
+	}
+}
+
+/// Information related to a child trie.
+/// This contains information needed to access a child trie
+/// content but also information needed to manage child trie
+/// from its parent trie (removal, move, update of root).
+///
+#[derive(PartialEq, Eq, Clone)]
+#[cfg_attr(feature = "std", derive(Debug, Hash, PartialOrd, Ord))]
+pub struct ChildTrie {
+	/// `KeySpace` for this child trie, see [`KeySpace`] for details.
+	keyspace: KeySpace,
+	/// Child trie current root, in case of modification
+	/// this is not updated.
+	root: Option<Vec<u8>>,
```

Comment:
How difficult would it be to make `ChildTrie` generic over `H: Hasher` and `root: Option<H::Out>`? It would save a lot of conversions in state machine and feels like the more appropriate type. Is this a bad idea?

Reply:
It shouldn't be very difficult, H being this trait everywhere (in fact it should be replace at my sense by more info: configuring the hasher is a thing, but I would like to configure a trie implementation, I remember starting such refacto and it is a lot of work (H is everywhere and using H as associated type of another 'every where trait' with more info did not behave super well, so it seems a bit involved unless if I just put another type aside H).
But for this change, it should be pretty straight forward.
The thing is that using Vec<u8> may have been a deliberate choice to avoid typing H::Out and have a easier implementation of child trie for different trie hash or trie implementation in child trie. This make sense if we consider using some dyn trait over trie, if we use the enum approach H would become an enum as H::Out so we can type it.
But heteregenous child_trie seems quite some work, so maybe having more solid ground on the current implementation may be better for now.
From memory, child trie still need change trie support, move implementation (should be easy after this pr), efficient deletion (pruning can be a bit tricky), and I may forge some.
One thing I would find really nice would be to make top trie a child trie and remove all duplication of api and caches (that is quite impacting and directly lead to using hierarchical child trie).
So to answer the question, I think it was made this way on purpose, so I simply keep it that way, changing it can be an idea (reverting to vec<u8> when needed).
But I would keep such change out of this PR (this PR does not change the previous behavior).